### DISCUSSION TITLE
Logic puzzle solution help

### DISCUSSION THREAD

**Discussion by emummel20** [uiua-lang/uiua#534] [General]
Hey no rush but I was wondering if someone could help with the solution to this puzzle in uiua. I have the solution in another lang and I would love to compare the two. This one example would help me a lot in terms of how to think in the paradigm.

Thanks

![image](https://github.com/user-attachments/assets/9ec296c9-f2a5-4f58-892f-24db7bafc7bc)

**Comment by dilr** [comment]
I just started learning Uiua a couple days ago, but I gave this a shot to help me learn the language and came up with some pretty good solutions. I don't know if the match how other people use the language, but I think at least the first solution goes with the grain of how the language wants you to use it.

Both solutions are brute force attempts which try every possible answer against all previous attempts. The first:

```uiua
SolvePasscode₃ ← |1 (
  # Generate all possible guesses
  ⟜(♭₂ ⇡ ↯:10 ⊣ △) # Input Guesses ? Input

  # Mask answering if each combination of guess
  # and input have exactly one matching digit.
  ⊸(= 1 ∵₁/+ ⊞=) # Mask Guesses ? Input Guesses

  # Return only guesses which are valid on all inputs
  ▽ /↧
)
```

I haven't been writing Uiua for that long, but I think this is probably about as clear as possible. (at least in terms of how the data flows through the program, there might be a better combination of primitives to use) Using `⍜now `, I find that it takes a bit over 5ms to run on my machine with the inputs given. I am in the habit of trying to get things to run faster after getting them working, so here is a faster solution:

```uiua
SolvePasscode₅ ← |1 (
  ⟜(♭₂ ⇡ ↯:10 ⊣ △) # Input Guesses ? Input

  ⊸(≡(=1 ∵₁/+= ¤) ⊙¤) # Mask Guesses ? Input Guesses
  ▽ /↧                # CorrectGuesses ? Mask Guesses
)
```
This uses `¤` inside and outside `≡` plus the pervasive `=` function to do the same thing as the `⊞` modifier, but it allows the sum reduction of the result before the entire mask is generated, which reduces peak memory usage. This runs in a bit more than 0.2ms on my machine. Its a bit harder to understand than the first solution, but it's nice to see such a minor modification can give an order of magnitude speedup.

As you can see by my naming scheme, I did try several other variations, but these are the most clear and fastest respectively.

**Reply by emummel20** [reply]
@dilr Thanks! Sorry I completely missed your response. Taking a look now. What argument do I pass SolvePasscode₃ or SolvePasscode₅ to get the result?

**Comment by amatgil** [comment]
Here's my shot at it:
```
Input ← [8_9_6 9_8_3 2_4_6 8_4_3]
Solve ← (
  ⊓⧅⋅⋅1¤ ⊙(⇡10) ⊣⊸△ # prep permutations
  ▽⤚≡(/×=₁ /+⍉ =¤)  # check that exactly one digit per row matches
)
Solve Input
```

[pad link](https://uiua.org/pad?src=0_15_0-dev_2__SW5wdXQg4oaQIFs4XzlfNiA5XzhfMyAyXzRfNiA4XzRfM10KU29sdmUg4oaQICgKICDiipPip4Xii4Xii4UxwqQg4oqZKOKHoTEwKSDiiqPiirjilrMgIyBwcmVwIHBlcm11dGF0aW9ucwogIOKWveKkmuKJoSgvw5c94oKBIC8r4o2JID3CpCkgICMgY2hlY2sgdGhhdCBleGFjdGx5IG9uZSBkaWdpdCBwZXIgcm93IG1hdGNoZXMKKQpTb2x2ZSBJbnB1dAo=)

**Reply by emummel20** [reply]
Checking this one too, thanks so much!

**Reply by emummel20** [reply]
Revisiting this - this is basically a brute force approach, correct? Is there a solution in uiua that's more efficient? Here's the solution in prolog

```
match([H|T],[H|T0]) :-
  maplist(dif,T,T0).
match([H|T],[H0|T0]) :-
  dif(H,H0),
  match(T,T0).

solution(X) :-
  match([8,9,6],X),
  match([9,8,3],X),
  match([2,4,6],X),
  match([8,4,3],X).
```

This solution uses nothing but the attempted numbers.

In your "prep permutations" step you generate set `[0 0 0]` which will never be used in the solution by the logic of the problem, which means we waste a lot of computation checking solutions we know to be impossible beforehand.
